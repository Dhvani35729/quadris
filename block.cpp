#include "block.h"

#include <iostream>
#include <vector>
#include <utility>

Block::Block(BlockType t, std::pair<int, int> pos, int level)
{
    this->type_ = t;
    this->coords_ = pos;
    this->level_ = level;
    aliveFor_ = 0;
    std::cout << "Block born" << std::endl;

    if (t == T_BLK)
    {
        this->mWidth_ = 3;
        this->mHeight_ = 3;
        matrix_.resize(this->mHeight_, std::vector<char>(this->mWidth_, ' '));
        matrix_[0][0] = 'T';
        matrix_[0][1] = 'T';
        matrix_[0][2] = 'T';
        matrix_[1][1] = 'T';
    }
};

Block::~Block()
{
    std::cout << "Block died" << std::endl;
};

std::vector<std::vector<char>> Block::getCells()
{
    return this->matrix_;
};

int Block::getBoxWidth()
{
    return this->mWidth_;
};

int Block::getBoxHeight()
{
    return this->mHeight_;
};

std::pair<int, int> Block::getPos()
{
    return this->coords_;
};

bool Block::moveBlock(Command)
{
    return true;
};
bool Block::rotateBlock(Command)
{
    return true;
};

void Block::dropBlock(){

};