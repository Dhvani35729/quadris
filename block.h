#ifndef BLOCK_H
#define BLOCK_H

#include "interpreter.h"

#include <vector>
#include <utility>

enum Type
{
    I_BLK,
    J_BLK,
    L_BLK,
    O_BLK,
    S_BLK,
    Z_BLK,
    T_BLK
};

// Block
class Block
{
public:
    Block(Type, std::pair<int, int>, int);
    ~Block();
    std::vector<std::vector<char>> getCells();
    std::pair<int, int> getBox();
    bool moveBlock(Command);
    bool rotateBlock(Command);
    void dropBlock();

private:
    int level_;
    Type type_;
    int aliveFor_;
    std::pair<int, int> coords_;
};

class HeavyBlock : public Block
{
public:
    HeavyBlock();
};

#endif