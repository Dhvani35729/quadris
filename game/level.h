#ifndef LEVEL_H
#define LEVEL_H

#include <string>
#include <memory>

#include "../blocks/block.h"

// Global seed for the randomness of block generation
extern int seed;

// Level - Strategy Pattern
// Level is an abstract class
class Level
{
public:
    // default constructor
    Level();

    // destructor
    virtual ~Level();

    // Pure virtual - nextBlock method (Strategy method)
    // Returns a new block to be played
    virtual std::shared_ptr<Block> nextBlock() = 0;

    // Takes in the number of blocks
    // placed since the last clear
    // Returns a special block to be played
    // By default, returns no blocks, but can be overriden
    virtual std::shared_ptr<Block> addSpecialBlock(int blockCount);

    // Set the file name to take the sequence from
    // Updates the block sequence vector
    virtual void setSequenceFile(std::string fileName);

    virtual void setIsRandom(bool r);

    // Getters
    int getLevelNum() const;

protected:
    // Helper method to get the next block
    // from the sequence
    BlockType getFromSequence();

    // Every level has a level number
    // Every level also has a bool representing
    // if it is random or not
    // Anytime a level is non random, it will
    // use the block sequence instead which
    // is generated from the file
    // NOTE: LevelZero is an exception to this,
    // but is only used for testing and thus we use this design.
    int levelNum_;
    std::string fileName_;

    std::vector<BlockType> blockSequence_;
    int curIndex_;

    bool isRandom_;
};

// LevelZero
// Derived from Level
// Note: Only used for testing purposes
class LevelZero : public Level
{
public:
    // constructor
    LevelZero(std::string fileName);

    // destructor
    ~LevelZero();

    // Concrete Strategy method
    std::shared_ptr<Block> nextBlock() override;

    // SetSequenceFile is overriden to an empty function
    // as this is a special level only for testing, and instead
    // uses the sequence.txt or provided -scriptfile
    void setSequenceFile(std::string) override;

private:
};

// LevelOne
// Derived from Level
class LevelOne : public Level
{
public:
    // default constructor
    LevelOne();

    // destructor
    ~LevelOne();

    // Concrete Strategy method
    std::shared_ptr<Block> nextBlock() override;
};

// LevelTwo
// Derived from Level
class LevelTwo : public Level
{
public:
    // default constructor
    LevelTwo();

    // destructor
    ~LevelTwo();

    // Concrete Strategy method
    std::shared_ptr<Block> nextBlock() override;
};

// LevelThree
// Derived from Level
class LevelThree : public Level
{
public:
    // default constructor
    LevelThree();

    // destructor
    ~LevelThree();

    // Concrete Strategy method
    std::shared_ptr<Block> nextBlock() override;
};

// LevelFour
// Derived from Level
class LevelFour : public Level
{
public:
    // default constructor
    LevelFour();

    // destructor
    ~LevelFour();

    // Concrete Strategy method
    std::shared_ptr<Block> nextBlock() override;

    // Returns a star block (1x1) every time a multiple of
    // 5 blocks are placed without clearing at least one row
    std::shared_ptr<Block> addSpecialBlock(int blockCount) override;
};

#endif