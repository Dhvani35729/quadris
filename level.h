#ifndef LEVEL_H
#define LEVEL_H

#include "block.h"

#include <string>
#include <memory>

// Level
class Level
{
public:
    virtual ~Level();
    virtual std::shared_ptr<Block> nextBlock() = 0;
    virtual std::shared_ptr<Block> addSpecialBlock(int);
    int getLevelNum() const;

protected:
    int levelNum_;
};

class LevelZero : public Level
{
public:
    LevelZero(std::string);
    ~LevelZero();
    std::shared_ptr<Block> nextBlock() override;

private:
    std::string fileName_;
    std::vector<char> blockSequence_;
    int curIndex;
};

class LevelOne : public Level
{
public:
    LevelOne();
    ~LevelOne();
    std::shared_ptr<Block> nextBlock() override;

private:
};

class LevelTwo : public Level
{
public:
    LevelTwo();
    ~LevelTwo();
    std::shared_ptr<Block> nextBlock() override;

private:
};

class LevelThree : public Level
{
public:
    LevelThree();
    ~LevelThree();
    std::shared_ptr<Block> nextBlock() override;

private:
    bool isRandom_;
};

class LevelFour : public Level
{
public:
    LevelFour();
    ~LevelFour();
    std::shared_ptr<Block> nextBlock() override;
    std::shared_ptr<Block> addSpecialBlock(int) override;

private:
    bool isRandom_;
};

#endif