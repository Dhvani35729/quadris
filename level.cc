#include "level.h"

#include <iostream>
#include <fstream>
#include <iostream>
#include <cstdlib>

using namespace std;

Level::Level()
{
    std::cout << "Level born" << std::endl;
    this->levelNum_ = -1;
}

Level::~Level()
{
    std::cout << "Level died" << std::endl;
}

LevelTwo::LevelTwo()
{
    std::cout << "LevelTwo born" << std::endl;
    this->levelNum_ = 2;
}

LevelZero::LevelZero(std::string f)
{
    std::cout << "LevelZero born" << std::endl;
    this->fileName_ = f;
    this->levelNum_ = 0;

    ifstream in(fileName_);

    string sequence(
        (std::istreambuf_iterator<char>(in)),
        std::istreambuf_iterator<char>());

    for (char &c : sequence)
    {
        if (c != ' ')
        {
            this->blockSequence_.push_back(c);
        }
    }

    this->curIndex = 0;
}

LevelZero::~LevelZero()
{
    std::cout << "LevelZero died" << std::endl;
}

LevelTwo::~LevelTwo()
{
    std::cout << "LevelTwo died" << std::endl;
}

int Level::getLevelNum() const
{
    return this->levelNum_;
}

Block *LevelZero::nextBlock()
{
    char nextBlock = this->blockSequence_[this->curIndex];
    BlockType type;
    if (nextBlock == 'I')
    {
        type = I_BLK;
    }
    else if (nextBlock == 'J')
    {
        type = J_BLK;
    }
    else if (nextBlock == 'L')
    {
        type = L_BLK;
    }
    else if (nextBlock == 'O')
    {
        type = O_BLK;
    }
    else if (nextBlock == 'S')
    {
        type = S_BLK;
    }
    else if (nextBlock == 'Z')
    {
        type = Z_BLK;
    }
    else if (nextBlock == 'T')
    {
        type = T_BLK;
    }
    else
    {
        type = BAD_BLK;
    }

    if (type != BAD_BLK)
    {
        Block *newBlock = new Block(type, std::make_pair(3, 0), this->levelNum_);
        this->curIndex += 1;

        if (this->curIndex == this->blockSequence_.size())
        {
            this->curIndex = 0;
        }
        return newBlock;
    }
    return nullptr;
}

Block *LevelTwo::nextBlock()
{
    int blkInd = rand() % BAD_BLK;
    cout << "Random block: " << blkInd << endl;
    BlockType type = (BlockType)blkInd;

    Block *newBlock = new Block(type, std::make_pair(3, 0), this->levelNum_);

    return newBlock;
}