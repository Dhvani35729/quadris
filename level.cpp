#include "level.h"

#include <iostream>
#include <fstream>
#include <iostream>
#include <cstdlib>
#include <memory>
#include <time.h>

// TODO: Do this.
// #define START_ROW 3
// #define START_COL 0

using namespace std;

std::shared_ptr<Block> Level::addSpecialBlock(int blockCount)
{
    return nullptr;
};

LevelZero::LevelZero(std::string f)
{
    std::cout << "LevelZero born" << std::endl;
    this->fileName_ = f;
    this->levelNum_ = 0;

    ifstream in(fileName_);

    string sequence(
        (std::istreambuf_iterator<char>(in)),
        std::istreambuf_iterator<char>());

    for (char &c : sequence)
    {
        if (c != ' ')
        {
            this->blockSequence_.push_back(c);
        }
    }

    this->curIndex = 0;
}

LevelOne::LevelOne()
{
    // Use current time as seed for random generator
    srand(time(0));
    std::cout << "LevelOne born" << std::endl;
    this->levelNum_ = 1;
}

LevelTwo::LevelTwo()
{
    // Use current time as seed for random generator
    srand(time(0));
    std::cout << "LevelTwo born" << std::endl;
    this->levelNum_ = 2;
}

LevelThree::LevelThree()
{
    // Use current time as seed for random generator
    srand(time(0));
    std::cout << "LevelThree born" << std::endl;
    this->levelNum_ = 3;
    this->isRandom_ = true;
}

LevelFour::LevelFour()
{
    // Use current time as seed for random generator
    srand(time(0));
    std::cout << "LevelFour born" << std::endl;
    this->levelNum_ = 4;
    this->isRandom_ = true;
}

Level::~Level()
{
    std::cout << "Level died" << std::endl;
}

LevelZero::~LevelZero()
{
    std::cout << "LevelOne died" << std::endl;
}

LevelOne::~LevelOne()
{
    std::cout << "LevelOne died" << std::endl;
}

LevelTwo::~LevelTwo()
{
    std::cout << "LevelTwo died" << std::endl;
}

LevelThree::~LevelThree()
{
    std::cout << "LevelThree died" << std::endl;
}

LevelFour::~LevelFour()
{
    std::cout << "LevelFour died" << std::endl;
}

int Level::getLevelNum() const
{
    return this->levelNum_;
}

std::shared_ptr<Block> LevelZero::nextBlock()
{
    char nextBlock = this->blockSequence_[this->curIndex];
    BlockType type;
    if (nextBlock == 'I')
    {
        type = I_BLK;
    }
    else if (nextBlock == 'J')
    {
        type = J_BLK;
    }
    else if (nextBlock == 'L')
    {
        type = L_BLK;
    }
    else if (nextBlock == 'O')
    {
        type = O_BLK;
    }
    else if (nextBlock == 'S')
    {
        type = S_BLK;
    }
    else if (nextBlock == 'Z')
    {
        type = Z_BLK;
    }
    else if (nextBlock == 'T')
    {
        type = T_BLK;
    }
    else
    {
        type = NUM_BLKS;
    }

    if (type != NUM_BLKS)
    {
        std::shared_ptr<Block> newBlock = std::make_shared<Block>(type, std::make_pair(3, 0), this->levelNum_, true);
        this->curIndex += 1;

        if (this->curIndex == this->blockSequence_.size())
        {
            this->curIndex = 0;
        }
        return newBlock;
    }
    return nullptr;
}

std::shared_ptr<Block> LevelOne::nextBlock()
{
    BlockType type;
    int blkInd = rand() % 6;
    if (blkInd == 5)
    {
        type = T_BLK;
    }
    else if (blkInd == 4)
    {
        blkInd = rand() % 2;
        blkInd = blkInd == 0 ? S_BLK : Z_BLK;
        type = (BlockType)blkInd;
    }
    else
    {
        type = (BlockType)blkInd;
    }

    std::shared_ptr<Block> newBlock = std::make_shared<Block>(type, std::make_pair(3, 0), this->levelNum_, true);

    return newBlock;
}

std::shared_ptr<Block> LevelTwo::nextBlock()
{

    int blkInd = rand() % NUM_BLKS;
    // cout << "Random block: " << blkInd << endl;
    BlockType type = (BlockType)blkInd;

    std::shared_ptr<Block> newBlock = std::make_shared<Block>(type, std::make_pair(3, 0), this->levelNum_, true);

    return newBlock;
}

std::shared_ptr<Block> LevelThree::nextBlock()
{
    BlockType type;
    int blkInd = rand() % 9;
    if (blkInd == 4)
    {
        type = T_BLK;
    }
    else if (blkInd == 5 || blkInd == 6)
    {
        type = S_BLK;
    }
    else if (blkInd == 7 || blkInd == 8)
    {
        type = Z_BLK;
    }
    else
    {
        type = (BlockType)blkInd;
    }

    std::shared_ptr<Block> newBlock = std::make_shared<HeavyBlock>(type, std::make_pair(3, 0), this->levelNum_, true);

    return newBlock;
}

std::shared_ptr<Block> LevelFour::nextBlock()
{
    BlockType type;
    int blkInd = rand() % 9;
    if (blkInd == 4)
    {
        type = T_BLK;
    }
    else if (blkInd == 5 || blkInd == 6)
    {
        type = S_BLK;
    }
    else if (blkInd == 7 || blkInd == 8)
    {
        type = Z_BLK;
    }
    else
    {
        type = (BlockType)blkInd;
    }

    std::shared_ptr<Block> newBlock = std::make_shared<HeavyBlock>(type, std::make_pair(3, 0), this->levelNum_, true);

    return newBlock;
}

std::shared_ptr<Block> LevelFour::addSpecialBlock(int blockCount)
{
    if (blockCount != 0 && blockCount % 5 == 0)
    {
        std::shared_ptr<Block> newBlock = std::make_shared<StarBlock>(std::make_pair(3, 5), this->levelNum_, false);
        return newBlock;
    }
    else
    {
        return nullptr;
    }
};
