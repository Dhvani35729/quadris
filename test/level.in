J
drop
right
O
drop
levelup
levelup
T
drop
S
clockwise
leveldown
J
drop
O
right
clockwise
right
clockwise
drop
T
counterclockwise
drop
S
levelup
drop
J
leveldown
drop