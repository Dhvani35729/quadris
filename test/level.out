Level:     0
Score:     0
Hi Score: 0
-----------
           
           
           
IIII       
           
           
           
           
           
           
           
           
           
           
           
           
           
           
-----------
Next:      
J  
JJJ


Level:     0
Score:     0
Hi Score: 0
-----------
           
           
           
J          
JJJ        
           
           
           
           
           
           
           
           
           
           
           
           
           
-----------
Next:      
J  
JJJ


Level:     0
Score:     0
Hi Score: 0
-----------
           
           
           
J          
JJJ        
           
           
           
           
           
           
           
           
           
           
           
J          
JJJ        
-----------
Next:      
OO 
OO 


Level:     0
Score:     0
Hi Score: 0
-----------
           
           
           
 J         
 JJJ       
           
           
           
           
           
           
           
           
           
           
           
J          
JJJ        
-----------
Next:      
OO 
OO 


Level:     0
Score:     0
Hi Score: 0
-----------
           
           
           
 OO        
 OO        
           
           
           
           
           
           
           
           
           
           
           
J          
JJJ        
-----------
Next:      
OO 
OO 


Level:     0
Score:     0
Hi Score: 0
-----------
           
           
           
OO         
OO         
           
           
           
           
           
           
           
           
           
           
 OO        
JOO        
JJJ        
-----------
Next:      
 SS
SS 


Level:     1
Score:     0
Hi Score: 0
-----------
           
           
           
OO         
OO         
           
           
           
           
           
           
           
           
           
           
 OO        
JOO        
JJJ        
-----------
Next:      
 SS
SS 


Level:     2
Score:     0
Hi Score: 0
-----------
           
           
           
OO         
OO         
           
           
           
           
           
           
           
           
           
           
 OO        
JOO        
JJJ        
-----------
Next:      
 SS
SS 


Level:     2
Score:     0
Hi Score: 0
-----------
           
           
           
TTT        
 T         
           
           
           
           
           
           
           
           
           
           
 OO        
JOO        
JJJ        
-----------
Next:      
 SS
SS 


Level:     2
Score:     0
Hi Score: 0
-----------
           
           
           
 SS        
SS         
           
           
           
           
           
           
           
           
TTT        
 T         
 OO        
JOO        
JJJ        
-----------
Next:      
  L
LLL


Level:     2
Score:     0
Hi Score: 0
-----------
           
           
           
 SS        
SS         
           
           
           
           
           
           
           
           
TTT        
 T         
 OO        
JOO        
JJJ        
-----------
Next:      
  L
LLL


Level:     2
Score:     0
Hi Score: 0
-----------
           
           
S          
SS         
 S         
           
           
           
           
           
           
           
           
TTT        
 T         
 OO        
JOO        
JJJ        
-----------
Next:      
  L
LLL


Level:     1
Score:     0
Hi Score: 0
-----------
           
           
S          
SS         
 S         
           
           
           
           
           
           
           
           
TTT        
 T         
 OO        
JOO        
JJJ        
-----------
Next:      
  L
LLL


Level:     1
Score:     0
Hi Score: 0
-----------
           
           
J          
JJJ        
           
           
           
           
           
           
           
           
           
TTT        
 T         
 OO        
JOO        
JJJ        
-----------
Next:      
  L
LLL


Level:     1
Score:     0
Hi Score: 0
-----------
           
           
           
  L        
LLL        
           
           
           
           
           
           
J          
JJJ        
TTT        
 T         
 OO        
JOO        
JJJ        
-----------
Next:      
ZZ 
 ZZ


Level:     1
Score:     0
Hi Score: 0
-----------
           
           
           
OO         
OO         
           
           
           
           
           
           
J          
JJJ        
TTT        
 T         
 OO        
JOO        
JJJ        
-----------
Next:      
ZZ 
 ZZ


Level:     1
Score:     0
Hi Score: 0
-----------
           
           
           
 OO        
 OO        
           
           
           
           
           
           
J          
JJJ        
TTT        
 T         
 OO        
JOO        
JJJ        
-----------
Next:      
ZZ 
 ZZ


Level:     1
Score:     0
Hi Score: 0
-----------
           
           
           
 OO        
 OO        
           
           
           
           
           
           
J          
JJJ        
TTT        
 T         
 OO        
JOO        
JJJ        
-----------
Next:      
ZZ 
 ZZ


Level:     1
Score:     0
Hi Score: 0
-----------
           
           
           
  OO       
  OO       
           
           
           
           
           
           
J          
JJJ        
TTT        
 T         
 OO        
JOO        
JJJ        
-----------
Next:      
ZZ 
 ZZ


Level:     1
Score:     0
Hi Score: 0
-----------
           
           
           
  OO       
  OO       
           
           
           
           
           
           
J          
JJJ        
TTT        
 T         
 OO        
JOO        
JJJ        
-----------
Next:      
ZZ 
 ZZ


Level:     1
Score:     0
Hi Score: 0
-----------
           
           
           
ZZ         
 ZZ        
           
           
           
           
           
  OO       
J OO       
JJJ        
TTT        
 T         
 OO        
JOO        
JJJ        
-----------
Next:      
TTT
 T 


Level:     1
Score:     0
Hi Score: 0
-----------
           
           
           
TTT        
 T         
           
           
           
           
           
  OO       
J OO       
JJJ        
TTT        
 T         
 OO        
JOO        
JJJ        
-----------
Next:      
TTT
 T 


Level:     1
Score:     0
Hi Score: 0
-----------
           
           
T          
TT         
T          
           
           
           
           
           
  OO       
J OO       
JJJ        
TTT        
 T         
 OO        
JOO        
JJJ        
-----------
Next:      
TTT
 T 


Level:     1
Score:     0
Hi Score: 0
-----------
           
           
           
TTT        
 T         
           
           
           
T          
TT         
T OO       
J OO       
JJJ        
TTT        
 T         
 OO        
JOO        
JJJ        
-----------
Next:      
OO 
OO 


Level:     1
Score:     0
Hi Score: 0
-----------
           
           
           
 SS        
SS         
           
           
           
T          
TT         
T OO       
J OO       
JJJ        
TTT        
 T         
 OO        
JOO        
JJJ        
-----------
Next:      
OO 
OO 


Level:     2
Score:     0
Hi Score: 0
-----------
           
           
           
 SS        
SS         
           
           
           
T          
TT         
T OO       
J OO       
JJJ        
TTT        
 T         
 OO        
JOO        
JJJ        
-----------
Next:      
OO 
OO 


Level:     2
Score:     0
Hi Score: 0
-----------
           
           
           
OO         
OO         
           
 SS        
SS         
T          
TT         
T OO       
J OO       
JJJ        
TTT        
 T         
 OO        
JOO        
JJJ        
-----------
Next:      
  L
LLL


Level:     2
Score:     0
Hi Score: 0
-----------
           
           
           
J          
JJJ        
           
 SS        
SS         
T          
TT         
T OO       
J OO       
JJJ        
TTT        
 T         
 OO        
JOO        
JJJ        
-----------
Next:      
  L
LLL


Level:     1
Score:     0
Hi Score: 0
-----------
           
           
           
J          
JJJ        
           
 SS        
SS         
T          
TT         
T OO       
J OO       
JJJ        
TTT        
 T         
 OO        
JOO        
JJJ        
-----------
Next:      
  L
LLL


Level:     1
Score:     0
Hi Score: 0
-----------
           
           
           
           
J          
JJJ        
 SS        
SS         
T          
TT         
T OO       
J OO       
JJJ        
TTT        
 T         
 OO        
JOO        
JJJ        
-----------
Next:      
  L
LLL